returns/contrib/hypothesis/laws.py

```diff
@@ -0,0 +1,151 @@
+
+    You can also pass different ``hypothesis`` settings inside:
+
+    .. code:: python
+
+      check_all_laws(IO, {'max_examples': 100})
+
+    Note:
+        Cannot be used inside doctests because of the magic we use inside.
+
+    See: https://mmhaskell.com/blog/2017/3/13/obey-the-type-laws
+    """
+    for interface, laws in container_type.laws().items():
+        for law in laws:
+            _create_law_test_case(
+                container_type,
+                interface,
+                law,
+                settings_kwargs=settings_kwargs,
+            )
+
+
+@contextmanager
+def container_strategies(container_type: Type[Lawful]) -> Iterator[None]:
+    """
+    Registers all types inside a container to resolve to a correct strategy.
+
+    For example, let's say we have ``Result`` type.
+    It is a subtype of ``ContainerN``, ``MappableN``, ``BindableN``, etc.
+    When we check this type, we need ``MappableN`` to resolve to ``Result``.
+
+    Can be used independently from other functions.
+    """
+    def factory(type_) -> st.SearchStrategy:
+        strategies: List[st.SearchStrategy[Any]] = []
+        if issubclass(container_type, ApplicativeN):
+            strategies.append(st.builds(container_type.from_value))
+        if issubclass(container_type, ResultLikeN):
+            strategies.append(st.builds(container_type.from_failure))
+        return st.one_of(*strategies)
+
+    # TODO: also register `KindN` instance of a type:
+    # TODO: or register `KindN` inside an `entrypoint`? By inspecting type
+    interfaces = {
+        base_type
+        for base_type in container_type.__mro__
+        if getattr(base_type, '__module__', '').startswith('returns.')
+    }
+    for interface in interfaces:
+        st.register_type_strategy(interface, factory)
+
+    yield
+
+    for interface in interfaces:
+        types._global_type_lookup.pop(interface)
+
+
+@contextmanager
+def pure_functions() -> Iterator[None]:
+    """
+    Context manager to resolve all ``Callable`` as pure functions.
+
+    It is not a default in ``hypothesis``.
+    """
+    def factory(thing) -> st.SearchStrategy:
+        like = (lambda: None) if len(
+            thing.__args__,
+        ) == 1 else (lambda *a, **k: None)
+
+        return st.functions(
+            like=like,
+            returns=st.from_type(thing.__args__[-1]),
+            pure=True,
+        )
+
+    previous = types._global_type_lookup[Callable]  # type: ignore
+    st.register_type_strategy(Callable, factory)  # type: ignore
+
+    yield
+
+    types._global_type_lookup.pop(Callable)  # type: ignore
+    st.register_type_strategy(Callable, previous)  # type: ignore
+
+
+def _run_law(
+    container_type: Type[Lawful],
+    law: Law,
+) -> Callable[[st.DataObject], None]:
+    def factory(source: st.DataObject) -> None:
+        with pure_functions():
+            with container_strategies(container_type):
+                source.draw(st.builds(law.definition))
```

Comment:
```suggestion
        with pure_functions(), container_strategies(container_type):
            source.draw(st.builds(law.definition))
```

See [here](https://docs.python.org/3/reference/compound_stmts.html#with), or just run `shed --refactor` :wink:

Reply:
We actually use an opposite rule: 

> WPS316 Found context manager with too many assignments

https://wemake-python-stylegui.de/en/latest/pages/usage/violations/consistency.html#wemake_python_styleguide.violations.consistency.MultipleContextManagerAssignmentsViolation